### ISSUE TITLE
SWI.py incompatible with py.test?

### ISSUE THREAD

**Issue by Seanny123** [nengo/nengo_gui#498]
@tcstewar the SWI.py module is giving me trouble. Basically, I'm trying to start a new thread for every test case I'm running with pytest (since they each use a new `nengo_gui.GUI` instance) and because of your use of Signal in SWI.py, it crashes.

Here is the exact error message:
```
Traceback (most recent call last):
  File "/home/sean/GitHub-linux/nengo_gui/nengo_gui/tests/conftest.py", line 46, in do_gui
    gui.start(port=port, browser=False)
  File "/home/sean/GitHub-linux/nengo_gui/nengo_gui/gui.py", line 86, in start
    addr=addr)
  File "/home/sean/GitHub-linux/nengo_gui/nengo_gui/swi.py", line 395, in start
    cls.begin_lifecycle(server, asynch)
  File "/home/sean/GitHub-linux/nengo_gui/nengo_gui/swi.py", line 435, in begin_lifecycle
    signal.signal(signal.SIGINT, cls._confirm_shutdown)
ValueError: signal only works in main thread
```

It does not occur if I just try creating a new thread like so:
```
import nengo_gui
import threading
import traceback as tb
import logging

logging.basicConfig(filename='example.log',level=logging.DEBUG)
port = 8080
model_path= "model_basic.py"

def do_gui():
    logging.info("Launching nengo_gui (port=%d)", port)
    #print("Launching nengo_gui (port=%d)", port)
    gui = nengo_gui.GUI(model_path)
    try:
        gui.start(port=port, browser=False)
    except:
        logging.debug("Server raised exception, suppressed:\n%s",
                      tb.format_exc(()))

# start the server in it's own thread
threading.Thread(target=do_gui).start()
```

Can that segment of SWI.py be changed? Am I doing the wrong approach with pytest?

**Comment by tcstewar** [comment]
I don't think I understand what the problem is.  You've given an example of code that does work; could you also include an example of code that doesn't work?  How was this running before when people were doing the selenium tests?  What has changed since then?  Did the signal thing exist back then?

**Comment by Seanny123** [comment]
Here is the (minimal and significantly reduced) [code that doesn't work](https://gist.github.com/Seanny123/9fd2a9d54d7bf78a7b72).

The test code was written during the first code sprint by Aaron and Peter B, so I don't know whether the signal thing existed back then or not.

Let me know if you need more details. The input I was expecting from you was an indication if this really was a SWI.py problem or not.

**Comment by tcstewar** [comment]
> Let me know if you need more details. The input I was expecting from you was an indication if this really was a SWI.py problem or not.

Ah, sorry, I think I misinterpretted what you were looking for here.  

According to the blame logs, that signal stuff was added to help with shutdown confirmation: https://github.com/nengo/nengo_gui/commit/fa9c6ef5b8cfafe33f7cb3483ef97026500e923a

However, there is an if statement right above the line that is giving you the error that checks to see if we are in "interactive" mode (i.e. are we running the server normally, as opposed to running it in some alternate situation, like inside IPython).  Have you tried setting interactive to false?  That should mean all that signal stuff isn't used, and I believe that's the correct way to indicate to the system that it's being run in a non-standard way.
